client/rest/build.gradle

```diff
@@ -19,6 +19,29 @@
 
 import org.elasticsearch.gradle.precommit.PrecommitTasks
 
+/**
+ * The rest client is a shaded jar. It contains the source of the rest client, as well as all the dependencies,
+ * shaded to the `internal.` package. 3 artifacts come out of this build process.
+ *
+ * 1) The original jar contents, using the `internal.` packages, but without the `internal.` code also in the jar.
+ *    The reason this jar was not overwritten was because gradle knows what its inputs and outputs are for a given task
+ *    and we should not alter that by overwriting the original jar. What we do is add a `nodeps` classifier to it instead.
+ *    This jar is removed from the output artifacts in cli, and retained in IDE builds.
+ * 2) A jar that contains *only* the `internal.` dependencies. This is a jar that is built before the src is compiled.
```

Comment:
I like this comment but I'd rephrase it, maybe moving the number 2 jar to the number 1 jar and explaining that it is also there for IDEs. I think it'd be good also to explain that we use the shaded package names in the source so we still jar up the bits that come out of javac without any post processing.